core/src/main/java/com/google/zxing/qrcode/encoder/MinimalEncoder.java

```diff
@@ -0,0 +1,778 @@
+ * We hypothetically assume that a mode is encoded using 1 bit (instead of 4) and a character is encoded in BYTE mode using 2 bit (instead of 8). Using these values we now attempt to encode the 
+ * four characters "1234".
+ * If we furthermore assume that in Version 1-9 the length field has 1 bit length so that it can encode up to 2 characters and that in Version 10-26 it has 2 bits length so that we can encode up
+ * to 2 characters then it is more efficient to encode with Version 10-26 than with Version 1-9 as shown below:
+ *
+ * Number of length bits small version (1-9): 1
+ * Number of length bits large version (10-26): 2
+ * Number of bits per mode item: 1
+ * Number of bits per character item: 2
+ * BYTE(1,2),BYTE(3,4): 1+1+2+2,1+1+2+2=12 bits
+ * BYTE(1,2,3,4): 1+2+2+2+2+2          =11 bits
+ *
+ * If we however change the capacity of the large encoding from 2 bit to 4 bit so that it potentially can encode 16 items, then it is more efficient to encode using the small encoding
+ * as shown below:
+ *
+ * Number of length bits small version (1-9): 1
+ * Number of length bits large version (10-26): 4
+ * Number of bits per mode item: 1
+ * Number of bits per character item: 2
+ * BYTE(1,2),BYTE(3,4): 1+1+2+2,1+1+2+2=12 bits
+ * BYTE(1,2,3,4): 1+4+2+2+2+2          =13 bits
+ *
+ * But as mentioned, it is not clear to me if this can ever happen with the actual values.
+ *
+ * ECI switching:
+ *
+ * In multi language content the algorithm selects the most compact representation using ECI modes. For example the
+ * it is more compactly represented using one ECI to UTF-8 rather than two ECIs to ISO-8859-6 and ISO-8859-1 if the text contains more ASCII characters (since they are represented as
+ * one byte sequence) as opposed to the case where there are proportionally more Arabic characters that require two bytes in UTF-8 and only one in ISO-8859-6.
+ *
+ * @author Alex Geller
+ */
+final class MinimalEncoder {
+
+  static final boolean DEBUG = false;
+
+  private enum VersionSize {
+    SMALL("Version 1-9)"),
+    MEDIUM("Version 10-26)"),
+    LARGE("Version 27-40");
+
+    String description;
+
+    VersionSize(String description) {
+        this.description = description;
+    }
+
+    public String toString() {
+        return description;
+    }
+  }
+
+  private String stringToEncode;
+  private Version version = null;
+  private boolean isGS1 = false;
+  private CharsetEncoder[] encoders;
+    
+
+/**Encoding is optional (default ISO-8859-1) and version is optional (minimal version is computed if not specified*/
+  MinimalEncoder(String stringToEncode,Version version,boolean isGS1) throws WriterException {
+    this.stringToEncode = stringToEncode;
+    if (version != null) {
+      this.version = version;
+    }
+    this.isGS1 = isGS1;
+    CharsetEncoder[] isoEncoders = new CharsetEncoder[15];
```

Comment:
What's magical about 15 here?

Reply:
15 is the maximum number of ISO-8859 encoders (1 through 16 and 12 doesn't exist but I am not compacting in the array)